### ISSUE TITLE
Request: replace printf() with C++ output streams

### ISSUE THREAD

**Issue by wimrijnders** [mn416/QPULib#29]
The current code is replete with `printf()` statements, for example in `Lib/Target/Pretty.cpp`:

```C++
void pretty(Instr instr)
{
  switch (instr.tag) {
...
    case ALU:
      if (instr.ALU.cond.tag != ALWAYS) {
        printf("where ");
        pretty(instr.ALU.cond);
        printf(": ");
      }
      pretty(instr.ALU.dest);
      printf(" <-%s ", instr.ALU.setFlags ? "{sf}" : "");
      pretty(instr.ALU.op);
      printf("(");
      pretty(instr.ALU.srcA);
      printf(", ");
      pretty(instr.ALU.srcB);
      printf(")\n");
      return;
...
```

For more flexibility wrt output, it would be preferable to output to a stream instead of directly to `stdout`. The code would then look like:

```C++
void pretty(std::ostream &os,Instr instr)
{
  switch (instr.tag) {
...
    case ALU:
      if (instr.ALU.cond.tag != ALWAYS) {
        os << "where ";
        pretty(os, instr.ALU.cond);
        os << ": ";
      }
      pretty(os, instr.ALU.dest);
      os << " <-" << (instr.ALU.setFlags ? "{sf}" : "") << " ";
      pretty(os, instr.ALU.op);
      os << "(";
      pretty(os, instr.ALU.srcA);
      os << ", ";
      pretty(os, instr.ALU.srcB);
      os << ")\n";
      return;
...
```

The advantage of this is that the output is target-agnostic; how to further handle it is deferred to the caller.

As before, if you can agree with this, I would be happy to implement it throughout the code.

**Comment by wimrijnders** [comment]
Using output streams can be done one better by returning a value instead of passing in a stream instance. For example:

```C++
std::string pretty(Instr instr)
{
  std::ostringstream os;

  switch (instr.tag) {
...
    case ALU:
      if (instr.ALU.cond.tag != ALWAYS) {
        os << "where " << pretty(instr.ALU.cond) << ": ";
      }
      os << pretty(instr.ALU.dest)<< " <-" << (instr.ALU.setFlags ? "{sf}" : "") << " ";
      os << pretty(instr.ALU.op) << "(" << pretty(instr.ALU.srcA) << ", " << pretty(instr.ALU.srcB) << ")\n";
      return;
...

  return os.c_str();
}
```

**Comment by wimrijnders** [comment]
Of course, some care must be taken to translate the format strings properly, e.g.:

    printf("# %ld.%06lds\n", tvDiff.tv_sec, tvDiff.tv_usec);

The majority of `printf()`s, however, do not have format strings, so overall the conversion is straightforward.

**Comment by mn416** [comment]
I think the effort to move all `printf`s to streams is not worth it and could easily introduce bugs that could go unnoticed for quite a while.

Happy for `printf` to be replaced by `fprintf` though.

**Comment by wimrijnders** [comment]
Preventing bugs of this kind is the reason why unit tests exist. In fact, I added a point for this very concern.

It would work as follows: set up a unit test for some (or even all) example programs[I], checking that the source/target output is as expected. Then, replace the printf's with the ostream's. The unit tests will indicate if there are any issues with the output.

I believe this is doable. I personally prefer the ostream-construct, as may already be apparent. It's more flexible.

----
[I] Except for `AutoTest`, which outputs 18MB of random code, hence not checkable anyway. I've actually tried it.
